File: src/backend/modules/m5_production/schemas/inventory.py
"""
M5 생산/SCM — 재고 관리 Pydantic 스키마
"""
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field


# ── 창고 ──

class WarehouseCreate(BaseModel):
    """창고 생성"""
    code: str = Field(..., max_length=30, description="창고코드")
    name: str = Field(..., max_length=100, description="창고명")
    zone_type: str = Field(..., description="구역유형 (raw/wip/finished/defective)")
    description: Optional[str] = None


class WarehouseResponse(BaseModel):
    """창고 응답"""
    id: str
    code: str
    name: str
    zone_type: str
    description: Optional[str] = None
    is_active: bool = True


# ── 재고 현황 ──

class InventoryItem(BaseModel):
    """재고 항목"""
    id: str
    product_id: str
    product_name: Optional[str] = None
    product_code: Optional[str] = None
    warehouse_id: str
    warehouse_name: Optional[str] = None
    zone_type: Optional[str] = None
    quantity: float = 0
    unit_cost: float = 0
    status: str = "available"
    safety_stock: int = 0


# ── 입출고/이관/조정 ──

class InventoryReceive(BaseModel):
    """입고"""
    product_id: str = Field(..., description="품목 ID")
    warehouse_id: str = Field(..., description="입고 창고 ID")
    quantity: float = Field(..., gt=0, description="입고 수량")
    unit_cost: float = Field(0, ge=0, description="단가")
    notes: Optional[str] = None


class InventoryIssue(BaseModel):
    """출고"""
    product_id: str = Field(..., description="품목 ID")
    warehouse_id: str = Field(..., description="출고 창고 ID")
    quantity: float = Field(..., gt=0, description="출고 수량")
    notes: Optional[str] = None
    reference_type: Optional[str] = None
    reference_id: Optional[str] = None


class InventoryTransfer(BaseModel):
    """이관 (창고 간 이동)"""
    product_id: str = Field(..., description="품목 ID")
    from_warehouse_id: str = Field(..., description="출발 창고 ID")
    to_warehouse_id: str = Field(..., description="도착 창고 ID")
    quantity: float = Field(..., gt=0, description="이동 수량")
    notes: Optional[str] = None


class InventoryAdjust(BaseModel):
    """재고 조정 (절대값)"""
    product_id: str = Field(..., description="품목 ID")
    warehouse_id: str = Field(..., description="창고 ID")
    new_quantity: float = Field(..., ge=0, description="조정 후 수량")
    notes: str = Field(..., min_length=1, description="조정 사유 (필수)")


# ── 이동 이력 ──

class TransactionResponse(BaseModel):
    """재고 이동 이력 항목"""
    id: str
    product_name: Optional[str] = None
    product_code: Optional[str] = None
    from_warehouse: Optional[str] = None
    to_warehouse: Optional[str] = None
    quantity: float
    transaction_type: str
    reference_type: Optional[str] = None
    notes: Optional[str] = None
    created_at: Optional[datetime] = None
    created_by_name: Optional[str] = None


# ── 부족 재고 ──

class ShortageItem(BaseModel):
    """안전재고 미달 항목"""
    product_id: str
    product_name: Optional[str] = None
    product_code: Optional[str] = None
    current_qty: float = 0
    safety_stock: int = 0
    shortage_qty: float = 0

File: src/frontend/src/api/production/inventory.ts
/**
 * M5 생산/SCM — 재고 관리 API 호출 함수
 */
import api from '../client';

/* ── 타입 정의 ── */

export interface Warehouse {
  id: string;
  code: string;
  name: string;
  zone_type: string;
  description?: string;
  is_active: boolean;
}

export interface InventoryItem {
  id: string;
  product_id: string;
  product_name?: string;
  product_code?: string;
  warehouse_id: string;
  warehouse_name?: string;
  zone_type?: string;
  quantity: number;
  unit_cost: number;
  status: string;
  safety_stock: number;
}

export interface ShortageItem {
  product_id: string;
  product_name?: string;
  product_code?: string;
  current_qty: number;
  safety_stock: number;
  shortage_qty: number;
}

export interface OrderMaterialCheck {
  order_id: string;
  order_no: string;
  has_shortage: boolean;
  materials: {
    product_id: string;
    product_name: string;
    product_code: string;
    required_qty: number;
    current_qty: number;
    safety_stock: number;
    shortage_qty: number;
    is_shortage: boolean;
  }[];
}

/* ── API 호출 ── */

export async function listWarehouses() {
  const res = await api.get('/production/warehouses');
  return res.data.data as Warehouse[];
}

export async function listInventory(params?: {
  warehouse_id?: string;
  zone_type?: string;
  search?: string;
  page?: number;
  size?: number;
}) {
  const res = await api.get('/production/inventory', { params });
  return res.data.data;
}

export async function receiveInventory(data: {
  product_id: string;
  warehouse_id: string;
  quantity: number;
  unit_cost?: number;
  notes?: string;
}) {
  const res = await api.post('/production/inventory/receive', data);
  return res.data;
}

export async function issueInventory(data: {
  product_id: string;
  warehouse_id: string;
  quantity: number;
  notes?: string;
}) {
  const res = await api.post('/production/inventory/issue', data);
  return res.data;
}

export async function transferInventory(data: {
  product_id: string;
  from_warehouse_id: string;
  to_warehouse_id: string;
  quantity: number;
  notes?: string;
}) {
  const res = await api.post('/production/inventory/transfer', data);
  return res.data;
}

export async function adjustInventory(data: {
  product_id: string;
  warehouse_id: string;
  new_quantity: number;
  notes: string;
}) {
  const res = await api.post('/production/inventory/adjust', data);
  return res.data;
}

export async function listTransactions(params?: {
  product_id?: string;
  warehouse_id?: string;
  tx_type?: string;
  page?: number;
  size?: number;
}) {
  const res = await api.get('/production/inventory/transactions', { params });
  return res.data.data;
}

export async function getShortageList() {
  const res = await api.get('/production/inventory/shortage');
  return res.data.data as ShortageItem[];
}

export async function checkOrderMaterials(orderId: string) {
  const res = await api.post(`/production/inventory/check-order/${orderId}`);
  return res.data.data as OrderMaterialCheck;
}

/* ── 견적서 단계 재고/원자재 사전 체크 ── */

export interface QuotationMaterialItem {
  material_id: string;
  material_name: string;
  material_code: string;
  required_qty: number;
  current_stock: number;
  shortage_qty: number;
  is_shortage: boolean;
}

export interface QuotationCheckItem {
  product_id: string;
  product_name: string;
  product_code: string;
  requested_qty: number;
  finished_stock: number;
  finished_shortage: number;
  need_production: boolean;
  materials: QuotationMaterialItem[];
}

export interface QuotationCheckResult {
  items: QuotationCheckItem[];
  summary: {
    ready_to_ship: boolean;
    can_produce: boolean;
    need_purchase: boolean;
    shortage_materials: string[];
  };
}

export async function checkQuotationMaterials(
  lines: { product_id: string; quantity: number }[],
) {
  const res = await api.post('/production/inventory/quotation-check', { lines });
  return res.data.data as QuotationCheckResult;
}
